magento/pwa-studio
# [SPIKE] Module Replacement Plugin - Extensibility POC

## Description

Implement a Module Replacement system in `pwa-buildpack` that can be used to inject code into the bundle at build time. To prove the point, this PR replaces the venia logo with a goose logo and the loading component with the `BounceLoader` component from the [react-spinners](https://www.npmjs.com/package/react-spinners) package.

Both these changes are provided as extensions using a sample package called [sample-venia-loader-icon](https://github.com/revanth0212/sample-venia-loader-icon).

Extensions can also be added while creating a PWA App using the `create-pwa` utility provided as part of the PWA Studio package.

To call a package an extension, it has to expose `extensions.json` in its root structure which has the details that `buildpack` will need to replace modules from the storefront.

## Related Issue
Closes [PWA-209](https://jira.corp.magento.com/browse/PWA-209)

### Verification Steps

For an existing project

1. Go to PWA Studio root folder.
2. Install the extension using `yarn add sample-venia-loader-icon`
3. Build and run the app using `yarn build && yarn stage:venia` or `yarn watch:venia`.
4. When you open the app in the browser, you should see that the logo and the loading icon have changed.

For a new project

1. Create a new project using the `create-pwa` tool by running this command `DEBUG_PROJECT_CREATION=1 ./pwa-studio/packages/create-pwa/bin/create-pwa`. Setting `DEBUG_PROJECT_CREATION` to `1` will take bundles for `@magento` packages from local rather than from npm.
2. This will start the `inquirer` which will ask you questions about the sample storefront you are trying to create.
3. Once you reach the question about extensions, type `sample-venia-loader-icon`. That is the sample storefront extension that we trying to install in the app.
4. Once the `create-pwa` tool is done with its magic, you will have the app structure ready and you can navigate into that project and run `yarn build && yarn start` to build and start the app.
5. Go to the URL provided in the terminal and you should see the goose logo instead of the default venia logo. Also you should see the loading icon change to the `BounceLoader` from `sample-venia-loader-icon`.

## Screenshots / Screen Captures

![image](https://user-images.githubusercontent.com/35203638/69989687-c7407280-1509-11ea-8f31-3d3d7a12c92c.png)
Logo File downloaded is different in both versions.

![image](https://user-images.githubusercontent.com/35203638/69990148-bfcd9900-150a-11ea-9721-be02fa8860bd.png)
The logo shown on the loading screen is different.

![image](https://user-images.githubusercontent.com/35203638/69991360-1d62e500-150d-11ea-9dd4-9cad157dc87e.png)
The logo shown in the header section is different.

## Checklist
None. This is a SPIKE.
